File: translator.js
const fs = require('fs');
const path = require('path');
const axios = require('axios');
const pptxExtractor = require('pptx-extractor');
const officegen = require('officegen');

/**
 * Translates Japanese text to English using Azure Translator API
 * @param {string} text - Text to translate
 * @returns {Promise<string>} - Translated text
 */
async function translateText(text) {
  if (!text || text.trim() === '') {
    return text;
  }

  const apiKey = process.env.AZURE_TRANSLATOR_KEY;
  const region = process.env.AZURE_TRANSLATOR_REGION;
  
  if (!apiKey || !region) {
    throw new Error('Azure Translator API key or region not configured');
  }

  const endpoint = 'https://api.cognitive.microsofttranslator.com';
  const url = `${endpoint}/translate?api-version=3.0&from=ja&to=en`;

  try {
    const response = await axios({
      method: 'post',
      url: url,
      headers: {
        'Ocp-Apim-Subscription-Key': apiKey,
        'Ocp-Apim-Subscription-Region': region,
        'Content-Type': 'application/json'
      },
      data: [{ text }]
    });

    if (response.data && response.data.length > 0 && 
        response.data[0].translations && 
        response.data[0].translations.length > 0) {
      return response.data[0].translations[0].text;
    } else {
      throw new Error('Translation API returned unexpected response');
    }
  } catch (error) {
    console.error('Translation error:', error.response?.data || error.message);
    throw new Error(`Failed to translate text: ${error.message}`);
  }
}

/**
 * Translates content of a PowerPoint file from Japanese to English
 * @param {string} filePath - Path to the PowerPoint file
 * @returns {Promise<string>} - Path to the translated file
 */
async function translatePowerPoint(filePath) {
  try {
    // Extract content from PPTX
    const extractedData = await pptxExtractor.extract(filePath);
    
    // Create a new PowerPoint file
    const pptx = officegen('pptx');

    // Process each slide
    for (const slide of extractedData.slides) {
      // Create a new slide
      const newSlide = pptx.makeNewSlide();

      // Get all text elements from the slide
      const textElements = slide.textElements || [];
      
      // Translate and add text elements to the new slide
      for (const textElement of textElements) {
        if (textElement.text && textElement.text.trim()) {
          const translatedText = await translateText(textElement.text);
          
          // Add text to the slide (simplified - adjust coordinates and formatting as needed)
          newSlide.addText(translatedText, {
            x: textElement.position?.x || 0,
            y: textElement.position?.y || 0,
            font_size: textElement.fontSize || 18,
            color: textElement.color || '000000'
          });
        }
      }
      
      // Handle images if present (copying from original)
      if (slide.images && slide.images.length > 0) {
        for (const image of slide.images) {
          // Add images to the new slide (simplified)
          if (image.path) {
            newSlide.addImage({ 
              path: image.path,
              x: image.position?.x || 0,
              y: image.position?.y || 0,
              cx: image.size?.width || 300,
              cy: image.size?.height || 200
            });
          }
        }
      }
    }

    // Generate the translated PowerPoint file
    const fileName = path.basename(filePath, '.pptx');
    const translatedFilePath = path.join(
      path.dirname(filePath),
      `${fileName}_translated.pptx`
    );

    return new Promise((resolve, reject) => {
      const outputStream = fs.createWriteStream(translatedFilePath);
      
      // Handle events
      pptx.on('error', (err) => {
        console.error('Error generating PPTX:', err);
        reject(err);
      });
      
      outputStream.on('error', (err) => {
        console.error('Error writing file:', err);
        reject(err);
      });
      
      outputStream.on('close', () => {
        resolve(translatedFilePath);
      });
      
      // Generate the PPTX file
      pptx.generate(outputStream);
    });
  } catch (error) {
    console.error('Error processing PowerPoint file:', error);
    throw new Error(`Failed to translate PowerPoint: ${error.message}`);
  }
}

module.exports = {
  translatePowerPoint,
  translateText
};

File: translator.py
import os
import json
import requests
import uuid
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

def translate_text(text, source_lang='ja', target_lang='en', custom_dict=None):
    """
    Translate text using Azure Translator API
    
    Args:
        text (str): Text to translate
        source_lang (str): Source language code (default: 'ja')
        target_lang (str): Target language code (default: 'en')
        custom_dict (dict): Custom dictionary for terminology (default: None)
    
    Returns:
        str: Translated text
    """
    # If text is empty, return as is
    if not text or text.strip() == "":
        return text
    
    # Apply custom dictionary if provided
    if custom_dict:
        for key, value in custom_dict.items():
            if key in text:
                text = text.replace(key, value)
    
    # Get API key and region from environment variables
    api_key = os.getenv("AZURE_TRANSLATOR_KEY")
    region = os.getenv("AZURE_TRANSLATOR_REGION")
    
    if not api_key or not region:
        raise ValueError("Azure Translator API key or region not configured")
    
    # API endpoint and parameters
    endpoint = "https://api.cognitive.microsofttranslator.com"
    path = "/translate"
    constructed_url = endpoint + path
    
    params = {
        "api-version": "3.0",
        "from": source_lang,
        "to": target_lang
    }
    
    headers = {
        "Ocp-Apim-Subscription-Key": api_key,
        "Ocp-Apim-Subscription-Region": region,
        "Content-type": "application/json",
        "X-ClientTraceId": str(uuid.uuid4())
    }
    
    # Prepare body for the request
    body = [{
        "text": text
    }]
    
    # Make the API request
    try:
        response = requests.post(constructed_url, params=params, headers=headers, json=body)
        response.raise_for_status()  # Raise exception for HTTP errors
        
        # Parse the response
        result = response.json()
        
        # Extract translated text
        if result and len(result) > 0 and "translations" in result[0]:
            translated_text = result[0]["translations"][0]["text"]
            return translated_text
        else:
            raise ValueError("Translation API returned unexpected response")
            
    except requests.exceptions.RequestException as e:
        raise Exception(f"Translation request failed: {str(e)}")
    except (KeyError, IndexError, json.JSONDecodeError) as e:
        raise Exception(f"Error parsing translation response: {str(e)}")

def load_custom_dictionary(file_path):
    """
    Load custom dictionary from CSV or Excel file
    
    Args:
        file_path (str): Path to the CSV or Excel file
    
    Returns:
        dict: Dictionary with source terms as keys and translations as values
    """
    custom_dict = {}
    
    # Implement dictionary loading based on file type
    # This is a placeholder for the actual implementation
    
    return custom_dict